bitbox-bridge/src/web.rs

```diff
@@ -184,6 +184,67 @@ impl AllowedHosts {
     }
 }
 
+// On windows, we use a native system dialog to prompt the user.
+// The browser based prompt does not work as we can't launch a browser from the service (not allowed).
+//
+// We use WTSSendMessageA as officially recommended:
+//
+// See https://learn.microsoft.com/en-us/windows/win32/services/interactive-services
+// > You can use the following techniques to interact with the user from a service on all supported versions of Windows:
+// > Display a dialog box in the user's session using the [WTSSendMessage](https://learn.microsoft.com/en-us/windows/win32/api/wtsapi32/nf-wtsapi32-wtssendmessagea) function.
+#[cfg(target_os = "windows")]
+async fn user_confirm(
+    _confirm_state: Arc<ConfirmState>,
+    message: String,
+    _base_url: &str,
+) -> Result<bool, ()> {
+    use windows_sys::Win32::System::RemoteDesktop::{
+        WTSGetActiveConsoleSessionId, WTSSendMessageA, WTS_CURRENT_SERVER_HANDLE,
+    };
+
+    use windows_sys::Win32::UI::WindowsAndMessaging::{IDYES, MB_YESNO, MESSAGEBOX_RESULT};
+    let title_c = c"BitBoxBridge";
+    let message_c = std::ffi::CString::new(message.as_str()).map_err(|_| ())?;
+
+    let mut response: MESSAGEBOX_RESULT = 0;
+    let timeout: u32 = 60; // 60 seconds
+
+    unsafe {
+        // Need the active user session ID so the dialog is shown to the user.
+        let current_session = WTSGetActiveConsoleSessionId();
+        if current_session == 0xFFFFFFFF {
+            // See https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-wtsgetactiveconsolesessionid#return-value
+            return Err(());
+        }
+        let result = WTSSendMessageA(
+            WTS_CURRENT_SERVER_HANDLE,
+            current_session,
+            title_c.as_ptr() as *const u8,
+            title_c.count_bytes() as u32,
+            message_c.as_ptr() as *const u8,
+            message_c.as_c_str().count_bytes() as u32,
```

Comment:
Just wondered about encoding.

WTSSendMessageA uses ANSI encoding, while WTSSendMessageW uses UTF-16 encoding.


The message includes the `origin.host()` string, which afaik can only have ASCII chars, so I think for the current use this is okay.

Not sure if it's worth the effort trying to figure out how to convert it to UTF16 and switch to WTSSendMessageW :thinking: In the worst case the message box would show a borked string I think if there is an encoding mismatch.

Reply:
@NickeZ I did this now and it works better than expected. I tested with this modified msg including some random unicode chars:

 `format!("Allow {} to connect to your BitBox Ɯ ⓴ ▗ ◕ ⚲ ⛵ ✅ ?", host),` and the unicode symbols were actually all shown on my Win11:

![Screenshot_2024-08-29_12-14-39](https://github.com/user-attachments/assets/0abaa6cf-0fe3-4f07-8641-1950c370818b)

Reply:
Wow cool